imports/server/mediasoup.ts

```diff
@@ -0,0 +1,763 @@
+  // "producerclose", and if you call .close() yourself, no event is emitted.
+  // Using the observer API makes the bookkeeping much more reliable
+
+  onWorkerCreated(worker: types.Worker) {
+    worker.observer.on('newrouter', Meteor.bindEnvironment(this.onRouterCreated.bind(this)));
+  }
+
+  onRouterCreated(router: types.Router) {
+    router.observer.on('newtransport', Meteor.bindEnvironment(this.onTransportCreated.bind(this)));
+
+    router.observer.on('close', Meteor.bindEnvironment(() => {
+      Ansible.info('Router was shut down', { call: router.appData.call, router: router.id });
+      this.routers.delete(router.appData.call);
+      Routers.remove({ routerId: router.id });
+    }));
+
+    this.routers.set(router.appData.call, router);
+    Routers.insert({
+      call: router.appData.call,
+      createdServer: serverId,
+      routerId: router.id,
+      rtpCapabilities: JSON.stringify(router.rtpCapabilities),
+      createdBy: router.appData.createdBy,
+    });
+  }
+
+  onTransportCreated(transport: types.Transport) {
+    transport.observer.on('newproducer', Meteor.bindEnvironment(this.onProducerCreated.bind(this)));
+    transport.observer.on('newconsumer', Meteor.bindEnvironment(this.onConsumerCreated.bind(this)));
+
+    transport.observer.on('close', Meteor.bindEnvironment(() => {
+      this.transports.delete(`${transport.appData.peer}:${transport.appData.direction}`);
+      Transports.remove({ transportId: transport.id });
+      TransportStates.remove({ transportId: transport.id });
+    }));
+
+    if (!(transport instanceof types.WebRtcTransport)) {
+      Ansible.warn('Ignoring unexpected non-WebRTC transport', { call: transport.appData.call, peer: transport.appData.peer, transport: transport.id });
+      return;
+    }
+
+    transport.observer.on('icestatechange', Meteor.bindEnvironment((iceState: types.IceState) => {
+      TransportStates.upsert({
+        createdServer: serverId,
+        transportId: transport.id,
+      }, {
+        $set: {
+          iceState,
+          createdBy: transport.appData.createdBy,
+        },
+      });
+    }));
+    transport.observer.on('iceselectedtuplechange', Meteor.bindEnvironment((iceSelectedTuple?: types.TransportTuple) => {
+      TransportStates.upsert({
+        createdServer: serverId,
+        transportId: transport.id,
+      }, {
+        $set: {
+          iceSelectedTuple: iceSelectedTuple ? JSON.stringify(iceSelectedTuple) : undefined,
+          createdBy: transport.appData.createdBy,
+        },
+      });
+    }));
+    transport.observer.on('dtlsstatechange', Meteor.bindEnvironment((dtlsState: types.DtlsState) => {
+      TransportStates.upsert({
+        createdServer: serverId,
+        transportId: transport.id,
+      }, {
+        $set: {
+          dtlsState,
+          createdBy: transport.appData.createdBy,
+        },
+      });
+    }));
+
+    this.producers.forEach((producer) => {
```

Comment:
This wants to get filtered down to just the call

Reply:
Fixed in 1609128